### PR TITLE
Add a hook that is called for entries after install.

### PR DESCRIPTION
**Pull request by robshakir**
```
  * (M) rib/rib.go
  * (M) rib/rib_test.go
    - add new ResolvedEntryFn that is called after AddEntry with
      full context of the set of RIBs.
    - rename existing each -> PostChangeHook.
  * (M) server/server.go
    - rename post change hook.
```

### CODE REVIEW

File: rib/rib.go [hunk 1]
```diff
@@ -97,6 +110,11 @@ type RIB struct {
 	// same AddXXX methods can be used along with network instance the operation
 	// referred to. The map is keyed by the operation ID.
 	pendingEntries map[uint64]*pendingEntry
+
+	// resolvedEntryHook is a function that is called for all entries that
+	// can be fully resolved in the RIB. In the current implementation it
+	// is called only for IPv4 entries.
+	resolvedEntryHook ResolvedEntryFn
```
Comment (sthesayi) [inline rib/rib.go:117]: nit: is it conceptually only called for v4 or we are just doing it for v4 now? If latter then maybe we can call this v4ResolvedEntryHook?

File: rib/rib.go [hunk 2]
```diff
@@ -237,14 +255,22 @@ type pendingEntry struct {
...
 	for _, nir := range r.niRIB {
+		nir.mu.Lock()
 		nir.postChangeHook = fn
+		nir.mu.Unlock()
 	}
 }
 
+// SetResolvedEntryHook asssigns the supplied hook to all network instance RIBs within
+// the RIB structure.
+func (r *RIB) SetResolvedEntryHook(fn ResolvedEntryFn) {
```
Comment (sthesayi) [inline rib/rib.go:270]: nit: Same comment here. Should this be SetIPv4ResolvedEntryHook ?

File: rib/rib_test.go [hunk 3]
```diff
@@ -2855,6 +2858,156 @@ func TestAddNetworkInstance(t *testing.T) {
...
+				},
+			},
+		}}
+
+		for i, op := range ops {
+			op.Id = uint64(i)
+			op.Op = spb.AFTOperation_ADD
+
+			if _, _, err := r.AddEntry(defName, op); err != nil {
+				panic(fmt.Sprintf("cannot add entry %s, %v", prototext.Format(op), err))
```
Comment (sthesayi) [inline rib/rib_test.go:2896]: over here I think you can use t.Fatalf

**Review by sthesayi — APPROVED** [review]

File: rib/rib.go (hunk 1 above)
Comment (robshakir) [reply rib/rib.go:117]: We are just doing it for v4 right now, it should be used for IPv6, MPLS and PBR going forward - hence keeping the name a bit more general.

File: rib/rib.go (hunk 2 above)
Comment (robshakir) [reply rib/rib.go:270]: Same comment as above.

File: rib/rib_test.go (hunk 3 above)
Comment (robshakir) [reply rib/rib_test.go]: Fixed -- starting a don't panic PR too.

**Pull request merged**